### PR TITLE
shell: devmem: address `cmd_dump` multiple call problem

### PR DESCRIPTION
**Pull request by ndrs-pst**
- correct number of `mand` and `opt` of `cmd_dump`
- address `cmd_dump` multiple call problem

### CODE REVIEW

**Review by jakub-uC — DISMISSED** [review]

File: subsys/shell/modules/devmem_service.c [hunk 1]
```diff
@@ -19,6 +19,8 @@
 #include <zephyr/shell/shell.h>
 #include <zephyr/sys/byteorder.h>
 
+extern void getopt_init(void);
```
Comment (ycsin) [inline subsys/shell/modules/devmem_service.c:22]: why not include `getopt.h`?

Comment (ndrs-pst) [reply subsys/shell/modules/devmem_service.c:22]: I can explain as follows: it seems that `getopt.h` will be included when `CONFIG_SHELL_GETOPT` is enabled.
However, using `getopt` only requires `CONFIG_GETOPT`. Therefore, if we change the configuration to `CONFIG_GETOPT`, there will be no need to directly extern `void getopt_init(void)`.

https://github.com/zephyrproject-rtos/zephyr/blob/28555ae9f4be81f0a1f3ba20e278a8594b372409/include/zephyr/shell/shell.h#L21-L23

Next step if we can expose <getopt.h> by `CONFIG_GETOPT` then we can utilize the following approach instead. 

```
struct getopt_state *getopt_state_get(void);

state = getopt_state_get();
state->optind = 1;
```

Comment (ndrs-pst) [reply subsys/shell/modules/devmem_service.c:22]: Oh by the way, it seems cannot pass CI test with (native_sim)

Comment (ndrs-pst) [reply subsys/shell/modules/devmem_service.c:23]: I've now managed to get the CI test to pass with the compile switch. The root cause stems from the incompatible implementation between `CONFIG_NATIVE_LIBC`'s `getopt` and `lib\posix\options\getopt\getopt.c`.
So, what could be the long-term solution for this?

Comment (cfriedt) [reply subsys/shell/modules/devmem_service.c:23]: One possible (preferable?) long term solution should be to use TLS for `optind`, etc. That way, threads will not clobber one another.

**Review by cfriedt — APPROVED** [review]
LGTM assuming CI is green

**Review by jakub-uC — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Comment by cfriedt** [comment]
@stephanosio - any chance you could kick CI somehow here? I think just merging would probably also be fine

**Comment by RICCIARDI-Adrien** [comment]
Hi, I've just encountered the same issue when using the `shell_module` example with Zephyr 0cb828d44376a07e79f9bf969f354e0c72972d98.
I did not test this fix, but I'm assuming it is solving the issue, so it would be nice to merge it.
Thanks !

**Comment by RICCIARDI-Adrien** [comment]
I get a similar problem with the CRC shell. Maybe using the thread local storage to keep a optind variable per thread, as suggested by @cfriedt, would be better to fix all this kind of issues.

**Comment by ycsin** [comment]
Trying to retrigger CI

**Pull request merged**